OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+  /// function is called on an empty DataTable_, nrow must be provided.
+  /// Otherwise, nrow is ignored. To add just one col, use addRow().
+  ///
+  /// \param first Beginning of range covered by the iterator.
+  /// \param last End of the range covered by the iterator.
+  /// \param nrow Number of rows to create in the DataTable_. This is only
+  ///             used(and required) when the function is called on an empty 
+  ///             DataTable_. Ignore otherwise.
+  /// \param allow_missing Allow for missing values. When set to false, this
+  ///                      function will throw if the input iterator fills up 
+  ///                      the last col only partially. When set to true, 
+  ///                      missing elements are set to SimTK::NaN.
+  ///
+  /// \throws OpenSim::ZeroElements If the number of elements produced by the
+  ///                               input iterator is zero.
+  /// \throws OpenSim::InvalidEntry If the function is called on an empty
+  ///                               DataTable_ without providing the argument
+  ///                               nrow.
+  /// \throws OpenSim::InvalidEntry If allow_missing is false and the input
+  ///                               iterator does not fill up the last col
+  ///                               completely.
+  template<typename InputIt>
+  void addCols(InputIt first, 
+               InputIt last, 
+               const size_t nrow = std::numeric_limits<size_t>::max(),
+               const bool allow_missing = false) {
+    if(!(first != last))
+      throw ZeroElements{"Input iterators produce zero elements."};
+    if((data.nrow() == 0 || data.ncol() == 0) && 
+       (nrow == std::numeric_limits<size_t>::max() || nrow == 0))
+      throw InvalidEntry{"DataTable is empty. 'nrow' argument must be" 
+                         " provided and it cannot be zero."};
+
+    data.resizeKeep(data.nrow() == 0 ? static_cast<int>(nrow) : data.nrow(), 
+                    data.ncol() + 1);
+    int row{0};
+    int col{data.ncol() - 1};
+    while(first != last) {
+      data.set(row, col, *first);
+      ++first; ++row;
+      if(row == static_cast<int>(data.nrow()) && first != last) {
+        row = 0;
+        ++col;
+        data.resizeKeep(data.nrow(), data.ncol() + 1);
+      }
+    }
+    if(!allow_missing && row != data.nrow())
+      throw InvalidEntry{"Input iterator did not produce enough elements to "
+                         "fill the last col."};
+  }
+
+  /// Bind another DataTable_ to this DataTable_ by row. The new elements will 
+  /// appear as the last rows of this DataTable_. Only data will be appended.
+  /// Metadata is not added. Columns retain their labels. To create a new 
+  /// DataTable_  that is a bind of two existing DataTable_(s), see 
+  /// rbindDataTables().
+  ///
+  /// \throws OpenSim::InvalidEntry If input DataTable_ has incorrect number of
```

Comment:
Maybe this is heresy, but don't think it is a good idea to attempt to document `throws`. The reason is that having the documentation implies that the list is exhaustive, but we have no way to ensure that. What is thrown could be anything from this method or from any method it invokes or anything thrown by the C++ runtime. Because we have no automated way of checking this, the documentation is more likely to be wrong than right (if not now, then later after some modifications have been made to low-level invoked methods). We also don't attempt to do this anywhere else in Simbody or OpenSim.

The C++ `throw` stuff has been mostly useless since it was invented. In C++ 11 the standards committee invented a new keyword `nothrow` to replace the previously-useless `throw()`. I do think `nothrow` is useful but it can only be used for very simple methods where you can determine for sure that an exception can't be thrown. This is useful because it affects whether move constructors will be used; std::vector at least requires nothrow move constructors or it will use copy constructors instead.